src/pip/_internal/operations/prepare.py

```diff
@@ -547,24 +547,29 @@ def _prepare_linked_requirement(self, req, parallel_builds):
             dist = _get_prepared_distribution(
                 req, self.req_tracker, self.finder, self.build_isolation,
             )
+        return dist
 
-            if download_dir:
-                if link.is_existing_dir():
-                    logger.info('Link is a directory, ignoring download_dir')
-                elif local_file:
-                    download_location = os.path.join(
-                        download_dir, link.filename
-                    )
-                    if not os.path.exists(download_location):
-                        shutil.copy(local_file.path, download_location)
-                        download_path = display_path(download_location)
-                        logger.info('Saved %s', download_path)
+    def save_linked_requirement(self, req):
```

Comment:
I find it particularly difficult to follow what's happening within this method. Can we simplify the conditionals, and make use of early returns here?

Reply:
I'll look into this soon-ish and see what I can do (I need to figure out when `req.archive` is expected, but my mind is not into it at this very moment :smile:).

Reply:
I dug a little deeper and unwind the confusions that made my head hurt in fab7966e73f9ed7e617ce7076f55d4f0d0f9c028 and 4acb5287e6c28cb5307568e07a134f20f50df98c.  Does the current look better now?